Spack and pip deployment

LTrestka opened the pull request:
* Refactored code to support pip packaging
* Created setup.py, and bin/ferry-cli - which runs as an executable
* Made pull request on spack scd_recipes repo: https://github.com/marcmengel/scd_recipes/pull/1
* Updated readme with full set of instructions, from setup, to authorization, to running
* Removed config.toml, and all references to it

To install with pip:
``` bash
pip install git+https://github.com/fermitools/Ferry-CLI.git
# Or
# clone and cd into the repo, then run:
pip install . 
ln -s /path/to/Ferry-CLI/bin/ferry-cli /usr/bin/ferry-cli # may need sudo depending on where you want it
```

To install with spack, get latest scd_recipes, then:
``` bash
spack install ferry-cli
spack load ferry-cli
```

LTrestka committed:
Added version info

LTrestka committed:
Merge branch 'filtered_lists' into spack_deployment

LTrestka committed:
Update README.md

LTrestka committed:
fixed version and copyright year

LTrestka committed:
Merge branch 'spack_deployment' of https://github.com/fermitools/Ferry-CLI into spack_deployment

LTrestka committed:
added setup.py, updated requirements and version

LTrestka committed:
I actually ended up refactoring the code a little bit, such that all of the files are now in a package folder: ferry_cli
Other notable changes include:
ferry.py is now located at ferry_cli.__main__.py
There are two sets of imports in a try/except statement like they are in the fife_wrap:
try:
    from ferry_cli.helpers.auth import Auth
    from ferry_cli.config import DIR
except ImportError:
    from helpers.auth import Auth
    from config import DIR
This was done because the pip installer requires each folder to essentially be a module (unless im doing it wrong), and submodules are defined with the full path, whereas running python3 ferry.py
only cares about the relative path.
DIR is used for file refs (config.toml, swagger.json), so that it doesnt look in the current directory when calling the executable.
With these changes, running python3 ferry.py ARGS will no longer work.
Instead, run: python3 ferry_cli ARGS
Additionally, from the root directory (Ferry-CLI), where the setup.py file is, you can run:
python3 -m pip install .
This will install the cli tool in your python packages, and generate a binary: ferry-cli
Once that is done, you can just run: ferry-cli ARGS  from anywhere
Once this is published, you can simply install the current tool by running:
pip install git+https://github.com/fermitools/Ferry-CLI.git@spack_deployment
Note that this is just where this change will be for now, it is not the spack package itself - that will be pushed to the spack scd repo

LTrestka committed:
Update README.md

LTrestka committed:
Updated permissions: chmod +x all .py files

LTrestka committed:
Merge branch 'spack_deployment' of https://github.com/fermitools/Ferry-CLI into spack_deployment

LTrestka committed:
Update setup.py

LTrestka committed:
Created non-compiled binary rather than entry point. This should work well with spack.

removed config.toml file, TConfig class, toml requirement, and cofig.toml references.

LTrestka committed:
Merge branch 'spack_deployment' of https://github.com/fermitools/Ferry-CLI into spack_deployment

LTrestka committed:
Cleaned up with pre-commit

LTrestka committed:
toml not required

LTrestka committed:
Added .spack/package.py here for safe storage

shreyb reviewed (changes requested):
Fantastic work so far on this.  There are a number of small issues (and a couple of larger ones) that I've indicated here, so please take your time going through those and ask me questions if any of the comments are unclear.

I haven't actually tested your instructions yet, but I'll go ahead and do that as we get closer to merging this.

LTrestka committed:
- Gave it an MIT license.
- Addressed comments in pr
  - Change DIR to CONFIG_DIR
  - corrected innacuracies in readme

LTrestka committed:
Repo access not necessary if public

shreyb reviewed (commented):
I think this is almost there.  Could you please resolve the merge conflict?

LTrestka committed:
Merge branch 'master' of https://github.com/fermitools/Ferry-CLI into spack_deployment
